web/src/lib/components/pages/record/profile-actions.tsx

```diff
@@ -0,0 +1,84 @@
+import { h, Component } from 'preact';
+import Profile from '../profile/profile';
+import User from '../../../user';
+
+interface Props {
+  user: User;
+}
+
+interface State {
+  profileFormVisible: boolean;
+  whyProfileVisible: boolean;
+}
+
+export default class ProfileActions extends Component<Props, State> {
+  state: State = {
+    profileFormVisible: false,
+    whyProfileVisible: false,
+  };
+
+  private toggleProfileForm = () => {
+    this.setState({ profileFormVisible: !this.state.profileFormVisible });
+  };
+
+  private toggleWhyProfile = () => {
+    this.setState({ whyProfileVisible: !this.state.whyProfileVisible });
+  };
+
+  render() {
+    return (
+      <div id="profile-actions">
+        <hr />
+        {this.state.profileFormVisible ? (
+          <div id="profile-form-container">
+            <a class="cancel" onClick={this.toggleProfileForm}>
+              Cancel
+            </a>
+            <Profile active="" user={this.props.user} />
+          </div>
+        ) : this.props.user.hasEnteredInfo() ? (
+          <a onClick={this.toggleProfileForm}>Edit Profile</a>
+        ) : (
+          this.renderOnboarding()
+        )}
+      </div>
+    );
+  }
+
+  private renderOnboarding() {
+    const { profileFormVisible, whyProfileVisible } = this.state;
+    return [
+      !profileFormVisible && (
+        <button
+          type="button"
+          id="create-profile-button"
+          onClick={this.toggleProfileForm}>
+          Create a profile
+        </button>
+      ),
+      <div>
+        {whyProfileVisible ? (
```

Comment:
instead of switching at the top level of the why-profile box, i think we should use a common element structure for it, and switch visibility with CSS based on a class (or so).

but we can do that in a separate PR after this one

Reply:
Right, I've seen that kind of CSS switching in a couple of places. Imo that's not the React way and might also incur some perf cost. I probably shouldn't make perf statements without measuring, but my reasoning would be: By always rendering the whole tree, you make React's diffs more expensive (though I guess your counter would be that it's a tradeoff for less DOM manipulation).